File: Parser/AST/InterfaceDefinitionAstNode.cs
namespace Raccoonlang.Parser.AST;

using Tokenizer;

public class InterfaceDefinitionAstNode 
{
    private ModifiersAstNode? modifiers;
    private Token? name;
    private GenericTypesAstNode? genericTypes;
    private GenericTypeConstraintsAstNode? genericConstraints;
    private InterfaceBodyAstNode? bodyContainer;

    public static InterfaceDefinitionAstNode? TryParse(Parser parser)
    {
        Parser.ParserState state = parser.ShelfState();
        InterfaceDefinitionAstNode node = new InterfaceDefinitionAstNode();

        node.modifiers = ModifiersAstNode.Parse(parser);

        if (parser.Peek().Type != TokenType.INTERFACE) {
            parser.RestoreState(state);
            return null;
        }

        parser.Take();

        node.name = parser.Take(TokenType.IDENTIFIER);
        node.genericTypes = GenericTypesAstNode.TryParse(parser);
        node.genericConstraints = GenericTypeConstraintsAstNode.Parse(parser);
        node.bodyContainer = InterfaceBodyAstNode.Parse(parser);

        return node;
    }
}

File: _src/main/java/raccoonlang/parser/ast/InterfaceDefinitionAstNode.java
package raccoonlang.parser.ast;

import raccoonlang.parser.Parser;
import raccoonlang.tokenizer.Token;
import raccoonlang.tokenizer.TokenType;

public class InterfaceDefinitionAstNode {

    public ModifiersAstNode modifiersAstNode;
    public Token name;
    public GenericTypesAstNode genericTypesAstNode;
    public GenericTypeConstraintsAstNode genericTypeConstraintsAstNode;
    public InterfaceBodyAstNode interfaceBodyAstNode;

    public static InterfaceDefinitionAstNode tryParse(Parser parser) {
        Parser.ParserState parserState = parser.shelfState();
        InterfaceDefinitionAstNode interfaceDefinitionAstNode = new InterfaceDefinitionAstNode();

        interfaceDefinitionAstNode.modifiersAstNode = ModifiersAstNode.parse(parser);

        if(parser.peek().getTokenType() != TokenType.INTERFACE) {
            parser.restore(parserState);
            return null;
        }
        parser.take();

        interfaceDefinitionAstNode.name = parser.take(TokenType.IDENTIFIER);
        interfaceDefinitionAstNode.genericTypesAstNode = GenericTypesAstNode.tryParse(parser);
        interfaceDefinitionAstNode.genericTypeConstraintsAstNode = GenericTypeConstraintsAstNode.parse(parser);
        interfaceDefinitionAstNode.interfaceBodyAstNode = InterfaceBodyAstNode.parse(parser);

        return interfaceDefinitionAstNode;
    }
}
